#' Cell filtering plot for scATAC-seq analysis
#'
#' Filtering the cells and non-cells based on number of UMIs and covered genes in each cell.
#'
#' @docType methods
#' @name ATACFilteringPlot
#' @rdname ATACFilteringPlot
#'
#' @param filepath Path of the scATAC-seq mapping summary file generated by \code{MAESTRO}.
#' @param name Name for the output cell filtering plot and valid cells file.
#' @param platform Denote platform of the scATAC-seq data, available options are "microfluidic",
#' "10x-genomics","sci-ATAC-seq". Default platform is "10x-genomics".
#' @param reads.cutoff Reads cutoff. Cells with less than \code{reads.cutoff} reads will be classified as non-cells.
#' Default is 1000.
#' @param frip.cutoff Fraction of reads in promoter cutoff. Cells with frip score less than \code{frip.cutoff} will be 
#' classified as non-cells. Default is 0.2, which means 20 percent reads should be in promoter regions.
#'
#' @author Chenfei Wang, Dongqing Sun
#'
#' @export

ATACFilteringPlot <- function(filepath, name, platform, reads.cutoff = 1000, frip.cutoff = 0.2)
{
  if(platform == "microfluidic" || platform == "sci-ATAC-seq"){
  mapping_matrix <- read.table(filepath, header = TRUE, sep = "\t", row.names = 1)
  png(paste0(name,"_scATAC_cell_filtering.png"),width=4.8,height=4.8, res = 300, units = "in")
  par(mai = c(0.85, 0.85, 0.25, 0.25))
  plot(log10(mapping_matrix[which(mapping_matrix[,5]<reads.cutoff|(mapping_matrix[,6]/mapping_matrix[,5])<frip.cutoff),5]+1),mapping_matrix[which(mapping_matrix[,5]<reads.cutoff|(mapping_matrix[,6]/mapping_matrix[,5])<frip.cutoff),6]/mapping_matrix[which(mapping_matrix[,5]<reads.cutoff|(mapping_matrix[,6]/mapping_matrix[,5])<frip.cutoff),5],
       xlim=c(0,5),ylim=c(0,1),pch='.',col='blue',ylab='Fraction of promoter reads',xlab='Reads passed filters (log10)')
       points(log10(mapping_matrix[which(mapping_matrix[,5]>=reads.cutoff&(mapping_matrix[,6]/mapping_matrix[,5])>=frip.cutoff),5]+1),mapping_matrix[which(mapping_matrix[,5]>=reads.cutoff&(mapping_matrix[,6]/mapping_matrix[,5])>=frip.cutoff),6]/mapping_matrix[which(mapping_matrix[,5]>=reads.cutoff&(mapping_matrix[,6]/mapping_matrix[,5])>=frip.cutoff),5],
       pch='.',col='red')
  legend("topright",c("cells","non-cells"),col=c("red","blue"),pch=20,bty = "n")
  dev.off()
  write.table(rownames(mapping_matrix[which(mapping_matrix[,5]>=reads.cutoff&(mapping_matrix[,6]/mapping_matrix[,5])>=frip.cutoff),]), 
              paste0(name,"_scATAC_validcells.txt"), sep = "\n", quote=F, row.names=F, col.names=F)
  }
  if(platform == "10x-genomics"){
  mapping_matrix <- read.csv(filepath)
  png(paste0(name,"_scATAC_cell_filtering.png"),width=4.8,height=4.8, res = 300, units = "in")
  par(mai = c(0.85, 0.85, 0.25, 0.25))
  plot(log10(mapping_matrix[which(mapping_matrix[,8]<reads.cutoff|(mapping_matrix[,14]/mapping_matrix[,8])<frip.cutoff),8]+1),mapping_matrix[which(mapping_matrix[,8]<reads.cutoff|(mapping_matrix[,14]/mapping_matrix[,8])<frip.cutoff),14]/mapping_matrix[which(mapping_matrix[,8]<reads.cutoff|(mapping_matrix[,14]/mapping_matrix[,8])<frip.cutoff),8],
       xlim=c(0,5),ylim=c(0,1),pch='.',col='blue',ylab='Fraction of promoter reads',xlab='Reads passed filters (log10)')
       points(log10(mapping_matrix[which(mapping_matrix[,8]>=reads.cutoff&(mapping_matrix[,14]/mapping_matrix[,8])>=frip.cutoff),8]+1),mapping_matrix[which(mapping_matrix[,8]>=reads.cutoff&(mapping_matrix[,14]/mapping_matrix[,8])>=frip.cutoff),14]/mapping_matrix[which(mapping_matrix[,8]>=reads.cutoff&(mapping_matrix[,14]/mapping_matrix[,8])>=frip.cutoff),8],
       pch='.',col='red')
  legend("topright",c("cells","non-cells"),col=c("red","blue"),pch=20,bty = "n")
  dev.off()
  write.table(as.character(mapping_matrix[which(mapping_matrix[,8]>=reads.cutoff&(mapping_matrix[,14]/mapping_matrix[,8])>=frip.cutoff),1]), 
              paste0(name,"_scATAC_validcells.txt"), sep = "\n", quote=F, row.names=F, col.names=F)
  }
}