#' Annotate scATAC-seq celltype based on marker genes
#'
#' Annotate the celltype for all clusters based on the marker genes and the input celltype signatures for scATAC-seq dataset.
#'
#' @docType methods
#' @name ATACAnnotateCelltype
#' @rdname ATACAnnotateCelltype
#'
#' @param ATAC Seurat object of clustered scATAC-seq dataset, generated by \code{\link{ATACRunSeurat}} function.
#' @param RPmatrix Data frame of regulatory potential matrix generated by MAESTRO. With genes as rows and cells as columns,
#' and gene RP score as values.
#' @param signatures Data frame of celltype signatures, with first column is celltype, and second column is
#' gene symbol. Check the immune signatures from CIBERSORT (Newman et al., Nature Method, 2015) in example for details.
#' @param min.score Minimum score required. For one cluster, if the score for all celltypes are less than 
#' \code{min.score}, the cluster will be annotated as "Others". Default is 0.1.
#' @param genes.test.use Denotes which test to use to identify genes. Default is "wilcox". 
#' @param genes.cutoff Identify differential expressed genes with adjusted p.value less than \code{genes.cutoff} as cluster speficic genes
#' for each cluster. Default cutoff is 1E-5.
#' @param orig.ident Orignal identity. If given, will use the original identity to annotate the celltypes. Default is NULL.
#'
#' @author Chenfei Wang, Dongqing Sun
#'
#' @return An ATAC Seurat object with celltype annotated, and the default assay have been set to gene activity.
#'
#'
#' @examples
#' data(pbmc.ATAC)
#' data(pbmc.RP)
#' data(human.immune.CIBERSORT)
#' pbmc.ATAC.res <- ATACRunSeurat(inputMat = pbmc.ATAC, project = "PBMC.scATAC.Seurat", method = "LSI")
#' pbmc.ATAC.res$ATAC <- ATACAnnotateCelltype(pbmc.ATAC.res$ATAC, pbmc.RP, human.immune.CIBERSORT, min.score = 0.1, genes.cutoff = 1E-3)
#' str(pbmc.ATAC.res$ATAC)
#'
#' @export

ATACAnnotateCelltype <- function(ATAC, RPmatrix, signatures, min.score = 0.1, genes.test.use = "wilcox", genes.cutoff = 1E-5, orig.ident = NULL)
{
  require(Seurat)
  require(ggplot2)
  RPmatrix <- RPmatrix[,colnames(ATAC)]
  ATAC[["ACTIVITY"]] <- CreateAssayObject(counts = RPmatrix)
  DefaultAssay(ATAC) <- "ACTIVITY"
  ATAC <- FindVariableFeatures(ATAC)
  ATAC <- NormalizeData(ATAC)
  ATAC <- ScaleData(ATAC)

  if(is.null(orig.ident)){
     message("Identify cluster specific genes based on RP score ...")
     cluster.genes <- NULL
     cluster.genes <- FindAllMarkersMAESTRO(object = ATAC, only.pos = TRUE, min.pct = 0.1, test.use = genes.test.use)
     cluster.genes <- cluster.genes[cluster.genes$p_val_adj<genes.cutoff, ]
     write.table(cluster.genes, paste0(ATAC@project.name, "_RPDiffGenes.tsv"), quote = F, sep = "\t")}
  
  ATAC <- RNAAnnotateCelltype(ATAC, cluster.genes, signatures, min.score = min.score, orig.ident = orig.ident)
  return(ATAC)
}