#' Identify differential features of single-cell transcriptome and regulome data for each of the clusters
#'
#' Identify differential features of single-cell data (genes for transcriptome and peaks for regulome) for each of the clusters
#'
#' @docType methods
#' @name FindAllMarkersMAESTRO
#' @rdname FindAllMarkersMAESTRO
#'
#' @param object Seurat object generated by \code{\link{RNARunSeurat}} function.
#' @param test.use Method to use to identify differential genes or peaks. Default is "wilcox", Wilcoxon Rank Sum test.
#' For scATAC-seq, "t" test is another option. For scRNA-seq, other available options are "bimod",
#' "roc", "t", "negbinom", "poisson", "LR", "MAST", "DESeq2", which supported by Seurat \code{link{FindAllMarkers}} function.
#' @param slot Slot to pull data from; note that if test.use is "negbinom", "poisson", or "DESeq2", 
#' slot will be set to "counts". Default is "data"
#' @param features Features to use. Default is to use all features.
#' @param min.pct Only test features that are detected in a minimum fraction of min.pct cells 
#' in either of the two populations. For genes, default is 0.1.
#' @param logfc.threshold Only test features which show X-fold difference (log-scale)
#' between two group of cells. For genes, default is 0.25.
#' @param latent.vars Variables to test, used only when test.use is one of 'LR', 'negbinom', 'poisson', or 'MAST'
#' @param min.cells.feature Minimum number of cells expressing the feature in at least one of the two groups, 
#' currently only used for poisson and negative binomial tests. Default is 3.
#' @param min.cells.group Minimum number of cells in one of the groups. Default is 3.
#' @param only.pos Only return positive markers, default is FALSE.
#' @param verbose Print a progress bar, default is TRUE.
#' @param return.thresh Only return markers that have a p-value < return.thresh, 
#' or a power > return.thresh (if the test is ROC)
#'
#' @author Dongqing Sun, Chenfei Wang
#'
#' @return A dataframe containing a ranked list of pytative markers and 
#' associated statics(p-value, ROC score, etc.)
#'
#' @export

FindAllMarkersMAESTRO <- function(object, test.use = 'wilcox', 
                                  features = NULL, min.pct = 0.1, logfc.threshold = 0.25,
                                  only.pos = FALSE, verbose = TRUE, return.thresh = 1e-2,
                                  min.cells.feature = 3, min.cells.group = 3,
                                  slot = "data", latent.vars = NULL){
  require(Seurat)
  ident.all <- sort(unique(Idents(object = object)))
  all.res <- lapply(0:(length(ident.all)-1), function(x){
    if (verbose) {
      message("Calculating cluster ",x)
    }
    test.res <-  tryCatch(expr = FindMarkersMAESTRO(object,ident.1 = x, ident.2 = NULL,test.use = 'wilcox', 
                                                    features = features, min.pct = min.pct, logfc.threshold = logfc.threshold,
                                                    only.pos = only.pos, verbose = verbose),
                          error = function(cond) {return(cond$message)})
    if(class(test.res) == "character"){
      message(paste0(test.res, ", no differential features identified"))
      test.res <- NULL
    }else{
      test.res$cluster <- as.character(x)
      test.res$gene <- rownames(test.res)
    }
    return(test.res)})
  all.res.df <- do.call("rbind", all.res)
  all.res.df <- all.res.df[which(all.res.df$p_val < return.thresh), ]
  return(all.res.df)
}

