#' Annotate scRNA-seq celltype based on marker genes
#'
#' Annotate the celltype for all clusters based on the marker genes and the input celltype signatures for scRNA-seq dataset.
#'
#' @docType methods
#' @name RNAAnnotateCelltype
#' @rdname RNAAnnotateCelltype
#'
#' @param RNA Seurat object of clustered scRNA-seq dataset, generated by \code{\link{RNARunSeurat}} function.
#' @param genes Data frame of differential expressed genes, generated by \code{\link{RNARunSeurat}} function.
#' @param signatures Data frame of celltype signatures, with first column is celltype, and second column is
#' gene symbol. Check the immune signatures from CIBERSORT (Newman et al., Nature Method, 2015) in example for details.
#' @param min.score Minimum score required. For one cluster, if the score for all celltypes are less than 
#' \code{min.score}, the cluster will be annotated as "Others". Default is 0.
#' @param orig.ident Orignal identity. If given, will use the original identity to annotate the celltypes. Default is NULL.
#'
#' @author Chenfei Wang
#'
#' @return A RNA Seurat object with annotated celltypes.
#'
#'
#' @examples
#' data(pbmc.RNA)
#' data(human.immune.CIBERSORT)
#' pbmc.RNA.res <- RNARunSeurat(inputMat = pbmc.RNA, project = "PBMC.scRNA.Seurat", min.c = 10, min.g = 100, dims.use = 1:15)
#' pbmc.RNA.res$RNA <- RNAAnnotateCelltype(pbmc.RNA.res$RNA, pbmc.RNA.res$genes, human.immune.CIBERSORT, min.score = 0.05)
#' str(pbmc.RNA.res$RNA)
#'
#' @importFrom Seurat DimPlot Idents
#' @importFrom ggplot2 ggsave
#' @export

RNAAnnotateCelltype <- function(RNA, genes, signatures, min.score = 0, orig.ident = NULL){
    if(is.null(orig.ident)){
        celltypes <- as.character(unique(signatures[,1]))
        signature_list <- sapply(1:length(celltypes),function(x){
                          return(toupper(as.character(signatures[which(signatures[,1]==celltypes[x]),2])))})
        names(signature_list) <- celltypes
            
        cluster_celltypes = sapply(as.integer(unique(RNA@meta.data$seurat_clusters))-1, function(x){
        idx = genes$cluster==x
        avglogFC = genes$avg_logFC[idx]
        names(avglogFC) = toupper(genes$gene[idx])
        score_cluster = sapply(signature_list, function(y){
                        score = sum(avglogFC[y], na.rm = TRUE) / log2(length(y))
                        return(score)
        })
        if(max(score_cluster, na.rm = TRUE)>min.score)
          cluster_celltypes = names(score_cluster)[which.max(score_cluster)]
        else
          cluster_celltypes = "Others"
        })
        
        current.cluster.ids = as.integer(unique(RNA@meta.data$seurat_clusters))-1
        new.cluster.ids = cluster_celltypes
        RNA@meta.data$assign.ident = Idents(RNA)[rownames(RNA@meta.data)]
        RNA@meta.data$assign.ident = plyr::mapvalues(x = RNA@meta.data$assign.ident,
                                                           from = current.cluster.ids, to = new.cluster.ids)
        p = DimPlot(object = RNA, label = TRUE, pt.size = 0.2, group.by = "assign.ident", label.size = 3, repel = T)
        ggsave(paste0(RNA@project.name, "_annotated.png"), p, width=6, height=4)}
    else{
        RNA$assign.ident <- orig.ident
        p = DimPlot(object = RNA, label = TRUE, pt.size = 0.2, group.by = "assign.ident", label.size = 3, repel = T)
        ggsave(file.path(paste0(RNA@project.name, "_original.png")), p,  width=5.5, height=4)}
    
    return(RNA)    
}